src/framework/audio/common/async/internal/channelimpl.h

```diff
@@ -0,0 +1,381 @@
+                // skip this thread
+                continue;
+            }
+
+            for (const Receiver* r : receiveThdata->receivers) {
+                if (!r->enabled.load(std::memory_order_acquire)) {
+                    continue;
+                }
+                CommMsg msg;
+                msg.receiver = r->receiver;
+                const auto& call = r->callback;
+                std::tuple<T...> vargs(args ...);
+                msg.func = [call, vargs]() {
+                    std::apply([call](const auto&... args) {
+                        call(args ...);
+                    }, vargs);
+                };
+                sendToQueue(sendThdata, receiveThdata->threadId, msg);
+            }
+        }
+    }
+
+    void sendDirect(const T&... args)
+    {
+        for (ThreadData* receiveThdata : m_threads) {
+            if (!receiveThdata) {
+                // there is no one further
+                break;
+            }
+
+            for (const Receiver* r : receiveThdata->receivers) {
+                if (!r->enabled.load(std::memory_order_acquire)) {
+                    continue;
+                }
+                r->callback(args ...);
+            }
+        }
+    }
+
+    void sendQueue(const T&... args)
+    {
+        const std::thread::id threadId = std::this_thread::get_id();
+
+        ThreadData& sendThdata = threadData(threadId);
+
+        for (ThreadData* receiveThdata : m_threads) {
+            if (!receiveThdata) {
+                // there is no one further
+                break;
+            }
+
+            for (const Receiver* r : receiveThdata->receivers) {
+                if (!r->enabled.load(std::memory_order_acquire)) {
+                    continue;
+                }
+                CommMsg msg;
+                msg.receiver = r->receiver;
+                const auto& call = r->callback;
+                std::tuple<T...> vargs(args ...);
+                msg.func = [call, vargs]() {
+                    std::apply([call](const auto&... args) {
+                        call(args ...);
+                    }, vargs);
+                };
+                sendToQueue(sendThdata, receiveThdata->threadId, msg);
+            }
+        }
+    }
+
+public:
+
+    ChannelImpl() = default;
+
+    ~ChannelImpl()
+    {
+        unregAllQueue();
+
+        for (ThreadData* thdata : m_threads) {
+            if (!thdata) {
+                break;
+            }
+
+            for (Receiver* r : thdata->receivers) {
+                if (r->receiver) {
+                    r->receiver->disconnectFromAsync(this);
+                }
+                delete r;
+            }
+
+            for (QueueData* qdata : thdata->queues) {
+                delete qdata;
+            }
+        }
+    }
+
+    // IConnectable
+    void disconnectAsync(Asyncable* a)
+    {
+        assert(a);
+
+        bool sending = isSending();
+        assert(sending);
```

Comment:
Was this supposed to be 
```suggestion
        assert(!sending);
```
?